pytorch_tabnet/tab_network.py

```diff
@@ -405,3 +376,71 @@ def forward(self, x):
         x = self.bn(x)
         out = torch.mul(x[:, :self.output_dim], torch.sigmoid(x[:, self.output_dim:]))
         return out
+
+
+class EmbeddingGenerator(torch.nn.Module):
+    """
+        Classical embeddings generator
+    """
+
+    def __init__(self, input_dim, cat_dims, cat_idxs, cat_emb_dim):
+        """ This is an embedding module for an entier set of features
+
+        Parameters
+        ----------
+        input_dim : int
+            Number of features coming as input (number of columns)
+        cat_dims : list of int
+            Number of modalities for each categorial features
+            If the list is empty, no embeddings will be done
+        cat_idxs : list of int
+            Positional index for each categorical features in inputs
+        cat_emb_dim : int or list of int
+            Embedding dimension for each categorical features
+            If int, the same embdeding dimension will be used for all categorical features
+        """
+        super(EmbeddingGenerator, self).__init__()
+        if cat_dims == []:
+            self.skip_embedding = True
+            self.post_embed_dim = input_dim
+        else:
+            self.skip_embedding = False
+            if isinstance(cat_emb_dim, int):
+                self.cat_emb_dims = [cat_emb_dim]*len(cat_idxs)
+                self.post_embed_dim = int(input_dim + (cat_emb_dim - 1)*len(cat_idxs))
+            else:
+                # check that all embeddings are provided
+                assert(len(cat_emb_dim) == len(cat_dims))
+                self.post_embed_dim = int(input_dim + np.sum(cat_emb_dim) - len(cat_emb_dim))
+                self.cat_emb_dims = cat_emb_dim
+
+            self.embeddings = torch.nn.ModuleList()
+            for cat_dim, emb_dim in zip(cat_dims, self.cat_emb_dims):
+                self.embeddings.append(torch.nn.Embedding(cat_dim, emb_dim))
+
+            # record continuous indices
+            self.continuous_idx = torch.ones(input_dim, dtype=torch.bool)
+            self.continuous_idx[cat_idxs] = 0
+
+    def forward(self, x):
+        """
+        Apply embdeddings to inputs
+        Inputs should be (batch_size, input_dim)
+        Outputs will be of size (batch_size, self.post_embed_dim)
+        """
+        if self.skip_embedding:
+            # no embeddings required
+            return x
+        else:
```

Comment:
Since you are returning in if, no need for else case

Reply:
done